packages/cardpay-reward-programs/cardpay_reward_programs/payment_tree.py

```diff
@@ -0,0 +1,130 @@
+#!/usr/bin/env python3
+
+import itertools
+from collections import defaultdict
+from copy import deepcopy
+from typing import Any, List, TypedDict
+
+import pyarrow as pa
+import sha3
+from eth_abi import encode_abi
+from eth_typing import ChecksumAddress
+from eth_utils import add_0x_prefix
+from merklelib import MerkleTree, beautify
+
+
+class Payment(TypedDict):
+    rewardProgramID: ChecksumAddress
+    payee: ChecksumAddress
+    paymentCycle: int
+    validFrom: int
+    validTo: int
+    token: ChecksumAddress
+    amount: int  # python should auto-detect big numbers
+
+
+def group_by(data_array, callback):
+    # remember: itertools.groupby requires that keys are sorted; it only groups common keys which are next to each other
+    sorted_data = sorted(data_array, key=callback)
+    return itertools.groupby(sorted_data, callback)
+
+
+class PaymentTree:
+    def __init__(self, payment_list: List[Payment], payment_cycle:int) -> None:
+        self.payment_nodes = self.aggregate_payments(payment_list)
+        self.data = [self.encode_payment(payment_node, payment_cycle) for payment_node in self.payment_nodes]
+        self.tree = MerkleTree(self.data, hashfunc)
+
+    def aggregate_payments(self, payments: Payment) -> List[Payment]:
```

Comment:
I copied this over, but I think it should just be down to the rules to create valid lists. I don't think anything breaks if there's more than one payment per person, and this function would fail when we start to consider non-claimable rewards (there's no amount). I think we can just delete this function.